### PR TITLE
chaos-daemon support graceful shutdown

### LINKED ISSUE #2650: subprocesses of chaos-daemon cannot gracefully stop
## Bug Report

**What version of Chaos Mesh are you using?**

v2.x.x

**What did you do?**
delete or upgrade chaos-daemon.

**What did you expect to see?**
all injected faults are recovered after chaos-daemon stops.

**What did you see instead?**
some faults (like http chaos) are not recovered.

**The Reason**
I found subprocesses of chaos-daemon are killed forcibly by the `runc`:

image
ps

`runc` sends SIGTERM to chaos-daemon, and the subprocesses (`nsexec` and `tproxy`) become orphans. Then, `runc` sends SIGKILL to `nsexec` and `tproxy` and they exit without graceful shutdown.

**Solutions**
We can handle SIGTERM in chaos-daemon, kill all subprocesses managed by the `bpm`, and wait for their shutdowns.

### PR DESCRIPTION
**Pull request by Hexilee**
### What problem does this PR solve?

 Close #2650

### Related changes

- [ ] Need to update `chaos-mesh/website`
- [ ] Need to update `Dashboard UI`
- Need to **cheery-pick to release branches**
  - [ ] release-2.1
  - [ ] release-2.0

### Checklist

Tests

- [ ] Unit test
- [x] E2E test
- [ ] No code
- [ ] Manual test (add steps below)

Side effects

- [ ] Breaking backward compatibility

### Release note 

```text
Please add a release note.

You can safely ignore this section if you don't think this PR needs a release note.
```

### DCO

If you find the DCO check fails, please run commands like below (Depends on the actual situations. For example, if the failed commit isn't the most recent) to fix it:

```shell
git commit --amend --signoff
git push --force
```

### CODE REVIEW

File: pkg/chaosdaemon/server.go [hunk 1]
```diff
@@ -147,7 +141,7 @@ func newGRPCServer(containerRuntime string, reg prometheus.Registerer, tlsConf t
 	s := grpc.NewServer(grpcOpts...)
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/chaosdaemon/server.go:141]: ignore

File: pkg/chaosdaemon/server.go [hunk 2]
```diff
@@ -159,45 +153,75 @@ type RegisterGatherer interface {
...
 		log.Error(err, "failed to create grpc server")
 		return err
 	}
 
-	g.Go(func() error {
+	go func() {
 		log.Info("Starting http endpoint", "address", httpBindAddr)
 		if err := httpServer.ListenAndServe(); err != nil {
 			log.Error(err, "failed to start http endpoint")
 			httpServer.Shutdown(context.Background())
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/chaosdaemon/server.go:200]: ignore

**Comment by Hexilee** [comment]
All tests are passed, PTAL! @YangKeao @STRRL @Andrewmatilde @cwen0

**Comment by Andrewmatilde** [comment]
In my opinion , graceful shitdown is just a death sig sender & waiter. So I am surprised about we have called delete() in deathsig map.  
I think their may have some problem in the design. 
I suggest :

1. For bpm , we don't actually need any channel to send any message to shutdown all the process. Bpm start every process , as a result bpm should manage them. I suggest bpm save every commands & send SIGTERM to them all by itself. You can add a waitgroup to count the death of all the process managed by bpm.
2. For gracefull shutdown, we need a simple moudle regist a death channel & send shutdown message to who regist the message.

**Comment by Hexilee** [comment]
> In my opinion , graceful shitdown is just a death sig sender & waiter. So I am surprised about we have called delete() in deathsig map. I think their may have some problem in the design. I suggest :
> 
>     1. For bpm , we don't actually need any channel to send any message to shutdown all the process. Bpm start every process , as a result bpm should manage them. I suggest bpm save every commands & send SIGTERM to them all by itself. You can add a waitgroup to count the death of all the process managed by bpm.
> 
>     2. For gracefull shutdown, we need a simple moudle regist a death channel & send shutdown message to who regist the message.

Actually, I'm going to refactor bpm, we can have another discussion, but it's irrelevant with this PR.

**Comment by Andrewmatilde** [comment]
I think we should have provided a common graceful shutdown tools , since the title is `chaos-daemon support graceful shutdown`. But now there are only a solution  for BPM .

**Comment by Hexilee** [comment]
> I think we should have provided a common graceful shutdown tools , since the title is `chaos-daemon support graceful shutdown`. But now there are only a solution for BPM .

We can just wait for closing other resources after chaos-daemon receiving SIGTERM, but I cannot figure out common solution to close any resource and wait for their exiting.

File: pkg/chaosdaemon/server.go [hunk 3]
```diff
@@ -159,45 +154,70 @@ type RegisterGatherer interface {
...
+	return &Server{
+		errors:       make(chan error),
+		daemonServer: daemonServer,
+		conf:         conf,
+		reg:          reg,
+	}, nil
+}
+
+// Start starts chaos-daemon.
+func (s *Server) Start() error {
```
Comment (STRRL) [inline pkg/chaosdaemon/server.go:182]: How about building `Start()` like `http.Server`'s `ListenAndServe()`? Do not use `Errors() <-chan error`.

File: pkg/chaosdaemon/server.go [hunk 4]
```diff
@@ -159,45 +154,70 @@ type RegisterGatherer interface {
...
+}
+
+// Errors return the error channel
+func (s *Server) Errors() <-chan error {
+	return s.errors
+}
 
-	return g.Wait()
+func (s *Server) Shutdown() {
+	s.daemonServer.backgroundProcessManager.Shutdown()
```
Comment (STRRL) [inline pkg/chaosdaemon/server.go:222]: we should also "Close" other resources like `grpc` and `http` server

**Review by STRRL — COMMENTED** [review]
Graceful shutdown usually relates to release/close opened resources. And it makes sense to` Shutdown()` return an error. Bu

File: pkg/chaosdaemon/server.go (hunk 4 above)
Comment (Hexilee) [reply pkg/chaosdaemon/server.go:222]: fine

File: pkg/chaosdaemon/server.go (hunk 3 above)
Comment (Hexilee) [reply pkg/chaosdaemon/server.go:182]: How about errgroup?

Comment (Hexilee) [reply pkg/chaosdaemon/server.go:187]: Fixed

File: pkg/chaosdaemon/server.go (hunk 4 above)
Comment (Hexilee) [reply pkg/chaosdaemon/server.go:220]: Fixed

**Review by Andrewmatilde — APPROVED** [review]
LGTM

**Review by YangKeao — APPROVED** [review]
LGTM

**Pull request merged**